### DISCUSSION TITLE
aurutils support

### DISCUSSION THREAD

**Discussion by VorpalBlade** [CyberShadow/aconfmgr#151] [Ideas]
I use [aurutils](https://github.com/AladW/aurutils) as my AUR helper of choice. I noticed that it is not listed as one of the one supported by aconfmgr. How difficult would it be adding support in aconfmgr for aurutils? Since you already support several different ones I presume it is relatively straightforward to add another one.

I'm currently trying out aconfmgr as an experiment, but if I stick long term with it I might even consider adding such support.

**Comment by CyberShadow** [comment]
How do you use aurutils? I use it too, and (at least to my understanding / the way I've set it up and use it) it's so fundamentally different from classic AUR helpers that I can't think of a straight-forward way to integrate it.

aurutils builds packages into a local repo, so using it with aconfmgr essentially involves just configuring pacman to use that repo. (Actually building packages multiple times on different machines seems against the spirit of aurutils anyway.) I also use some helpers + a hacky script  to build up the list of packages used in the configuration, to ensure the aurutils repo contains only packages that are actually being referenced and built (see https://github.com/AladW/aurutils/issues/840).

**Comment by VorpalBlade** [comment]
Hm, I do see the potential problem. What would be nice would be if aconfmgr could call out to aur to build required packages if they are missing from the repository.

This is especially useful when bootstrapping a new computer. I have yet to set up a way to sync `/home/custompkgs` between computers, and it does not contain exactly the same files (as different computers have different needs due to having different hardware) but that is an idea to keep in mind.

There is still a bootstrap problem though here: If I run `AddPackage` for something that doesn't exist yet because `/home/custompkgs` has yet to be set up on said computer. This would require running `aconfmgr` twice with special logic to deal with this, breaking the rule about `apply` being idempotent.

**Reply by CyberShadow** [reply]
I think the canonical way to use aurutils on multiple machines is to have it installed on exactly one of your machines, with the rest having e.g. `sftp://...` repos configured in `pacman.conf`.

There's a few workarounds for the bootstrap problem. The simplest is to ignore aconfmgr's aspiration to be idempotent and remember to run it twice, as you mentioned; `ConfigWarning` could be used to output a reminder at the end of the first run. Another is to ignore aconfmgr's aspiration towards a functional configuration, and add imperative commands directly to it (e.g. `sudo mkdir /home/custompkgs` in `~/.config/aconfmgr/10-setup.sh`).

If all that's needed is to install some files before installing any packages, there is `priority_files`, which is used to set up things like users or the pacman configuration.

If you think it's possible to add aurutils integration in some sufficiently-unopinionated way, I'd encourage you to attempt it. Adding an AUR helper is by itself not difficult, see e.g. https://github.com/CyberShadow/aconfmgr/pull/90/files.

**Reply by VorpalBlade** [reply]
This setup of server/client doesn't quite work for me, as I travel and use different computers. I seldom have both my laptop and my desktop running at the same time to save power as well. What would be more sensible for me would be to use peer to peer syncing of the repo when both are on. Possibly using syncthing (which I already use for other things).

I will have to think about how to go about possibly adding aurutils. I do indeed not see an obvious solution as of now.
